### PR TITLE
Add AWS Autoscaling Engine

### LINKED ISSUE #83: Add AWS AutoscalingEngine
#### Why should this feature be added?

Several users have expressed interest in using Cerebral in conjunction with AWS. Let's add an AWS AutoscalingEngine to achieve that.

#### How would you like the feature to work?

We need to define what an AWS engine can/should look like.

- Are users limited to using it in conjunction with an AWS ASG?
- What are the limitations for users who have provisioned Kubernetes using `kops`?
  - We can't overwrite the AWS ASG bounds because `kops` will pick up that change - we should be able to update the `desired` count though?

#### What alternatives have been considered?

N/A

#### Additional Context

N/A

### PR DESCRIPTION
**Pull request by mattkelly**
#### What does this pull request accomplish?

 Adds support for autoscaling in AWS.

 Documentation is currently light and a bit unclear IMO - this is something that I plan on tackling with a follow-up "getting started" documentation PR.

 #### What issue(s) does this fix?
* Closes #83

 #### Additional Considerations

 ### Testing

 #### Setup

 In my case:

 - `make`
 - `docker tag containership/cerebral:latest mattkelly/cerebral:latest`
 - `docker push mattkelly/cerebral:latest`

 #### Instructions

 I launched a cluster via `kops` and then ran Cerebral using the image above.

-----------------------------------

kops cluster:

```
apiVersion: kops/v1alpha2
kind: Cluster
metadata:
  creationTimestamp: 2019-02-06T18:52:11Z
  name: autoscaling-test.k8s.local
spec:
  api:
    loadBalancer:
      type: Public
  authorization:
    rbac: {}
  channel: stable
  cloudProvider: aws
  configBase: s3://kops-autoscaling/autoscaling-test.k8s.local
  etcdClusters:
  - etcdMembers:
    - instanceGroup: master
      name: a
    name: main
  - etcdMembers:
    - instanceGroup: master
      name: a
    name: events
  iam:
    allowContainerRegistry: true
    legacy: false
  kubelet:
    anonymousAuth: false
    authenticationTokenWebhook: true
    authorizationMode: Webhook
  kubernetesApiAccess:
  - 0.0.0.0/0
  kubernetesVersion: 1.11.5
  masterInternalName: api.internal.autoscaling-test.k8s.local
  masterPublicName: api.autoscaling-test.k8s.local
  networkCIDR: 172.20.0.0/16
  networking:
    kubenet: {}
  nonMasqueradeCIDR: 100.64.0.0/10
  sshAccess:
  - 0.0.0.0/0
  subnets:
  - cidr: 172.20.32.0/19
    name: us-east-1a
    type: Public
    zone: us-east-1a
  topology:
    dns:
      type: Public
    masters: public
    nodes: public
 ```

kops instance groups:

```
apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2019-02-07T02:55:10Z
  labels:
    kops.k8s.io/cluster: autoscaling-test.k8s.local
  name: master
spec:
  image: kope.io/k8s-1.11-debian-stretch-amd64-hvm-ebs-2018-08-17
  machineType: t2.medium
  maxSize: 1
  minSize: 1
  role: Master
  subnets:
  - us-east-1a
---
apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2019-02-06T18:55:15Z
  labels:
    kops.k8s.io/cluster: autoscaling-test.k8s.local
  name: nodes
spec:
  image: kope.io/k8s-1.11-debian-stretch-amd64-hvm-ebs-2018-08-17
  machineType: t2.small
  maxSize: 5
  minSize: 1
  nodeLabels:
    node-pool: workers
  role: Node
  subnets:
  - us-east-1a
```

---------------------------------

Cerebral Deployment with secret:

```
---
apiVersion: v1
kind: Secret
metadata:
  name: cluster-autoscaler-aws-engine
  namespace: kube-system
data:
  aws-access-key-id: <redacted>
  aws-secret-access-key: <redacted>
---
apiVersion: apps/v1
kind: Deployment
metadata:
  namespace: kube-system
  name: cerebral
  labels:
    app.kubernetes.io/name: cerebral
spec:
  replicas: 1
  selector:
    matchLabels:
      app.kubernetes.io/name: cerebral
  template:
    metadata:
      labels:
        app.kubernetes.io/name: cerebral
    spec:
      serviceAccountName: cluster-autoscaler
      tolerations:
      - effect: NoSchedule
        key: node-role.kubernetes.io/master
      nodeSelector:
        node-role.kubernetes.io/master: ""
      containers:
      - name: cerebral
        image: mattkelly/cerebral:latest
        imagePullPolicy: Always
        env:
        - name: LOG_LEVEL
          value: DEBUG
        - name: AWS_ACCESS_KEY_ID
          valueFrom:
            secretKeyRef:
              name: cluster-autoscaler-aws-engine
              key: aws-access-key-id
        - name: AWS_SECRET_ACCESS_KEY
          valueFrom:
            secretKeyRef:
              name: cluster-autoscaler-aws-engine
              key: aws-secret-access-key
```

Cerebral RBAC (this is modified from `cluster-autoscaler` just to get up and running - will clean up and adjust in next PR):

```
---
apiVersion: v1
kind: ServiceAccount
metadata:
  labels:
    k8s-addon: cluster-autoscaler.addons.k8s.io
    k8s-app: cluster-autoscaler
  name: cluster-autoscaler
  namespace: kube-system
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: cluster-autoscaler
  labels:
    k8s-addon: cluster-autoscaler.addons.k8s.io
    k8s-app: cluster-autoscaler
rules:
- apiGroups: [""]
  resources: ["events","endpoints"]
  verbs: ["create", "patch"]
- apiGroups: [""]
  resources: ["pods/eviction"]
  verbs: ["create"]
- apiGroups: [""]
  resources: ["pods/status"]
  verbs: ["update"]
- apiGroups: [""]
  resources: ["endpoints"]
  resourceNames: ["cluster-autoscaler"]
  verbs: ["get","update"]
- apiGroups: [""]
  resources: ["nodes"]
  verbs: ["watch","list","get","update"]
- apiGroups: [""]
  resources: ["pods","services","replicationcontrollers","persistentvolumeclaims","persistentvolumes"]
  verbs: ["watch","list","get"]
- apiGroups: ["extensions"]
  resources: ["replicasets","daemonsets"]
  verbs: ["watch","list","get"]
- apiGroups: ["policy"]
  resources: ["poddisruptionbudgets"]
  verbs: ["watch","list"]
- apiGroups: ["apps"]
  resources: ["statefulsets", "replicasets"]
  verbs: ["watch","list","get"]
- apiGroups: ["storage.k8s.io"]
  resources: ["storageclasses"]
  verbs: ["watch","list","get"]
- apiGroups: ["cerebral.containership.io"]
  resources: ["*"]
  verbs: ["*"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: cluster-autoscaler
  namespace: kube-system
  labels:
    k8s-addon: cluster-autoscaler.addons.k8s.io
    k8s-app: cluster-autoscaler
rules:
- apiGroups: [""]
  resources: ["configmaps"]
  verbs: ["create"]
- apiGroups: [""]
  resources: ["configmaps"]
  resourceNames: ["cluster-autoscaler-status"]
  verbs: ["delete","get","update"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: cluster-autoscaler
  labels:
    k8s-addon: cluster-autoscaler.addons.k8s.io
    k8s-app: cluster-autoscaler
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-autoscaler
subjects:
  - kind: ServiceAccount
    name: cluster-autoscaler
    namespace: kube-system
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: cluster-autoscaler
  namespace: kube-system
  labels:
    k8s-addon: cluster-autoscaler.addons.k8s.io
    k8s-app: cluster-autoscaler
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: cluster-autoscaler
subjects:
  - kind: ServiceAccount
    name: cluster-autoscaler
    namespace: kube-system
```

----------------------------

Cerebral CRs:

```
---
apiVersion: cerebral.containership.io/v1alpha1
kind: AutoscalingEngine
metadata:
  name: aws
spec:
  type: aws
---
apiVersion: cerebral.containership.io/v1alpha1
kind: AutoscalingGroup
metadata:
  name: kops-nodes-asg
spec:
  nodeSelector:
    "node-pool": "workers"
  policies:
  - kubernetes-cpu
  engine: aws
  suspended: false
  cooldownPeriod: 60
  maxNodes: 5
  minNodes: 1
---
apiVersion: cerebral.containership.io/v1alpha1
kind: AutoscalingPolicy
metadata:
  name: kubernetes-cpu
spec:
  metricsBackend: kubernetes
  metric: cpu_percent_allocation
  scalingPolicy:
    scaleUp:
      threshold: 15
      comparisonOperator: ">="
      adjustmentType: absolute
      adjustmentValue: 1
  pollInterval: 15
  samplePeriod: 5
---
apiVersion: cerebral.containership.io/v1alpha1
kind: MetricsBackend
metadata:
  name: kubernetes
spec:
  type: kubernetes
```

-------------------------

After applying the CRs, because of the low `scaleUp` `threshold`, a scale up was triggered:

Before:
```
➜ kc get nodes
NAME                            STATUS   ROLES    AGE   VERSION
ip-172-20-46-132.ec2.internal   Ready    node     7m    v1.11.5
ip-172-20-53-52.ec2.internal    Ready    master   1d    v1.11.5
```

Logs:
```
2019-02-09T15:16:51.507Z        DEBUG   Performing cpu allocation calculation of 5 pods across 1 nodes
2019-02-09T15:16:51.507Z        DEBUG   Poller for ASP "kubernetes-cpu" got value 64.000000
2019-02-09T15:16:51.507Z        DEBUG   ScaleManager: got scale request: {asgName:kops-nodes-asg direction:0 adjustmentType:0 adjustmentValue:1 ignoreCooldown:false errCh:0xc0006ca420}
2019-02-09T15:16:51.507Z        INFO    Event(v1.ObjectReference{Kind:"AutoscalingPolicy", Namespace:"", Name:"kubernetes-cpu", UID:"5b1d46ec-2c79-11e9-99b0-12aebbb782a4", APIVersion:"cerebral.containership.io/v1alpha1", ResourceVersion:"319705", FieldPath:""}): type: 'Normal' reason: 'ScaleUpAlerted' Alert triggered to scale up by 1.00 (absolute)
2019-02-09T15:16:51.507Z        DEBUG   Selected node ip-172-20-46-132.ec2.internal has providerID "aws:///us-east-1a/i-013412c66f8060cd3", instanceID "i-013412c66f8060cd3"
2019-02-09T15:16:51.827Z        INFO    AWS AutoscalingEngine aws is requesting AWS to scale to 2
2019-02-09T15:16:52.259Z        INFO    Event(v1.ObjectReference{Kind:"AutoscalingGroup", Namespace:"", Name:"kops-nodes-asg", UID:"f24b42a4-2c7b-11e9-99b0-12aebbb782a4", APIVersion:"cerebral.containership.io/v1alpha1”, ResourceVersion:"319972", FieldPath:""}): type: 'Normal' reason: 'ScaledUp' Scaled up to 2 nodes using strategy “”
```

After:
```
➜ kc get nodes
NAME                            STATUS   ROLES    AGE   VERSION
ip-172-20-33-243.ec2.internal   Ready    node     39s   v1.11.5
ip-172-20-46-132.ec2.internal   Ready    node     9m    v1.11.5
ip-172-20-53-52.ec2.internal    Ready    master   1d    v1.11.5
```

And we see the CPU allocation percentage cut in half:

```
2019-02-09T15:25:44.104Z        DEBUG   Performing cpu allocation calculation of 6 pods across 2 nodes
2019-02-09T15:25:44.104Z        DEBUG   Poller for ASP "kubernetes-cpu" got value 37.000000
```

 ### Dependencies
 * None

### CODE REVIEW

**Comment by mattkelly** [comment]
Example manifests will be part of #94 and a "Getting Started on AWS" document will be in a future PR.

File: README.md [hunk 1]
```diff
@@ -55,6 +55,7 @@ Support for a different `AutoscalingEngine` can be added by implementing the [en
 Because an `AutoscalingGroup` is defined by a label selector, the provider (or some other entity) must be able to label nodes when they are added.
 
 The currently available engines include:
+* [AWS][aws-engine]
```
Comment (ashleyschuett) [inline README.md]: 🎉

File: pkg/autoscaling/engines/aws/aws.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+		return nil, errors.New("name must be provided")
+	}
+
+	if nodeLister == nil {
+		return nil, errors.New("node lister must be provided")
+	}
+
+	// Note that aws-sdk-go pulls AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY
+	// directly from the environment
+	sess := session.New(aws.NewConfig().WithRegion(getRegion()))
```
Comment (ashleyschuett) [inline pkg/autoscaling/engines/aws/aws.go]: is there any way to check if this is valid before creating the client and attaching it to the engine?

File: pkg/autoscaling/engines/aws/aws.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+	providerID := selectedNode.Spec.ProviderID
+	if providerID == "" {
+		return false, errors.Errorf("selected node %s does not have providerID available", selectedNode.Name)
+	}
+
+	instanceID := instanceIDFromProviderID(providerID)
+
+	log.Debugf("Selected node %s has providerID %q, instanceID %q", selectedNode.Name, providerID, instanceID)
+
+	result, err := e.client.DescribeAutoScalingInstances(&awsautoscaling.DescribeAutoScalingInstancesInput{
```
Comment (ashleyschuett) [inline pkg/autoscaling/engines/aws/aws.go]: lets pull this and it's error checking out into its own function

File: pkg/autoscaling/engines/aws/aws.go (hunk 2 above)
Comment (mattkelly) [reply pkg/autoscaling/engines/aws/aws.go]: Not using this function but it looks like [`session.New()`](https://docs.aws.amazon.com/sdk-for-go/api/aws/session/#New) is deprecated anyway. Nice catch!

File: pkg/autoscaling/engines/aws/aws.go (hunk 3 above)
Comment (mattkelly) [reply pkg/autoscaling/engines/aws/aws.go]: Good idea, will do.

**Comment by mattkelly** [comment]
@ashleyschuett thanks for the thoughtful review. This is now updated and retested!

File: pkg/autoscaling/engines/aws/aws.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+		return nil, errors.New("name must be provided")
+	}
+
+	if nodeLister == nil {
+		return nil, errors.New("node lister must be provided")
+	}
+
+	// Note that aws-sdk-go pulls AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY
+	// directly from the environment
+	sess, err := session.NewSession(aws.NewConfig().WithRegion(getRegion()))
```
Comment (ashleyschuett) [inline pkg/autoscaling/engines/aws/aws.go]: Oh i should have asked the first time. How long does the session last for? Does it need to be refreshed?

Comment (mattkelly) [reply pkg/autoscaling/engines/aws/aws.go]: Good question. My understanding is that `NewConfig()` defaults to an `EnvProvider` for the credentials, and these credentials [never expire](https://docs.aws.amazon.com/sdk-for-go/api/aws/credentials/#EnvProvider).

**Pull request merged**